python/uv/private/lock.bzl

```diff
@@ -15,22 +15,258 @@
+            mandatory = True,
+            doc = """\
+The string to input for the 'uv pip compile'.
+""",
+        ),
+        "_template": attr.label(
+            default = "//python/uv/private:pip_compile_template",
+            cfg = "exec",
+            executable = True,
+            doc = """\
+The template to be used for 'uv pip compile'. This is either .ps1 or bash
+script depending on what the target platform is executed on.
+""",
+        ),
+    },
+    toolchains = [
+        _uv_toolchain,
+        _py_toolchain,
+    ],
+)
+
+def _lock_run_impl(ctx):
+    info = ctx.attr.lock[_LockInfo]
+    uv = info.uv
+    srcs = info.srcs
+    py_runtime = info.py
+
+    args = [
+        uv.short_path,
+        "pip",
+        "compile",
+    ] + info.args + [
+        src.short_path
+        for src in srcs
+    ] + [
+        "--python",
+        py_runtime.interpreter.short_path,
+    ]
+
+    executable = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.expand_template(
+        template = info.template,
+        substitutions = {
+            "replace": "{}/{}".format(ctx.label.package, ctx.attr.output),
+            "uv pip compile": " ".join(args),
+        },
+        output = executable,
+        is_executable = True,
+    )
+
+    runfiles = ctx.runfiles(
+        transitive_files = depset(
+            srcs + [uv],
+            transitive = [
+                py_runtime.files,
+            ],
+        ),
+    )
+    return [
+        DefaultInfo(
+            executable = executable,
+            runfiles = runfiles,
+        ),
+    ]
+
+_lock_run = rule(
+    implementation = _lock_run_impl,
+    doc = """\
+""",
+    attrs = {
+        "lock": attr.label(
+            doc = "The lock target that is doing locking in a build action.",
+            providers = [_LockInfo],
+        ),
+        "output": attr.string(
+            doc = "The output that we would be updated.",
+        ),
+    },
+    executable = True,
+)
+
+def _maybe_file(path):
+    """A small function to return a list of existing outputs.
+
+    If the file referenced by the input argument exists, then it will return
+    it, otherwise it will return an empty list. This is useful to for programs
+    like pip-compile which behave differently if the output file exists and
+    update the output file in place.
+
+    The API of the function ensures that path is not a glob itself.
+
+    Args:
+        path: {type}`str` the file name.
+    """
+    for p in native.glob([path], allow_empty = True):
+        if path == p:
+            return p
+
+    return None
+
+def lock(*, name, srcs, out, args = [], env = None, **kwargs):
     """Pin the requirements based on the src files.
 
     Differences with the current {obj}`compile_pip_requirements` rule:
-    - This is implemented in shell and `uv`.
+    - This is implemented as a rule that performs locking in a build action.
+    - Additionally one can use the runnable target.
```

Comment:
Please document the additional public targets with a `Generates public targets:` block, e.g.

```
Generates public targets:
  * `<name>` bla bla
  * `<name>.update` bla bla
```
```